lib/internal/streams/end-of-stream.js

```diff
@@ -63,22 +63,35 @@ function eos(stream, opts, callback) {
 
   const wState = stream._writableState;
   const rState = stream._readableState;
+  const state = wState || rState;
 
   const onlegacyfinish = () => {
     if (!stream.writable) onfinish();
   };
 
+  // TODO (ronag): Improve soft detection to include core modules and
+  // common ecosystem modules that do properly emit 'close' but fail
+  // this generic check.
+  const willEmitClose = (
+    state &&
+    state.autoDestroy &&
+    state.emitClose &&
+    state.closed === false
+  );
+
   let writableFinished = stream.writableFinished ||
     (wState && wState.finished);
   const onfinish = () => {
     writableFinished = true;
+    if (willEmitClose && (!stream.readable || readable)) return;
```

Comment:
`!stream.readable || readable´ is for the case where the stream is `Duplex`, but we might not want to wait for `'end'` in which case `autoDestroy` won't be called an `'close'` won't be emitted. An example of this is:

```js
await pipeline(readable, duplex)
```